### PR TITLE
libsecret - store passwords in key chain.

### PR DESCRIPTION
**Pull request by snaunton**
#31 

Store passwords in keychain, using libsecret, rather than in plaintext.

@TingPing - I understand your points from #31 re sandboxing/disk encryption, but, lots of people still use distros with libsecret and --without-libsecret removes support if you want to build without it.

### CODE REVIEW

File: src/trg-client.c [hunk 1]
```diff
@@ -309,7 +318,36 @@ int trg_client_populate_with_settings(TrgClient * tc)
...
+#ifdef HAVE_LIBSECRET
+    GError *error = NULL;
+
+    gchar *hostname = trg_prefs_get_string(tc->priv->prefs, TRG_PREFS_KEY_HOSTNAME, TRG_PREFS_PROFILE);
+    gint64 port = trg_prefs_get_int(tc->priv->prefs, TRG_PREFS_KEY_PORT, TRG_PREFS_PROFILE);
+    gchar *rpc_url_path = trg_prefs_get_string(tc->priv->prefs, TRG_PREFS_KEY_RPC_URL_PATH, TRG_PREFS_PROFILE);
+    gboolean ssl = trg_prefs_get_bool(tc->priv->prefs, TRG_PREFS_KEY_SSL, TRG_PREFS_PROFILE);
+    gchar *username = trg_prefs_get_string(tc->priv->prefs, TRG_PREFS_KEY_USERNAME, TRG_PREFS_PROFILE);
+
+    gchar *password = secret_password_lookup_sync (TRG_SECRET_SCHEMA, NULL, &error,
```
Comment (TingPing) [inline src/trg-client.c]: Blocking IO on the main thread.

File: src/trg-client.c [hunk 2]
```diff
@@ -316,38 +321,60 @@ int trg_client_populate_with_settings(TrgClient * tc)
...
     g_free(hostname);
     g_free(rpc_url_path);
     g_free(username);
 
-    return password;
-#else
-    return tc->priv->password;
+    gint64 timeout = g_get_monotonic_time () + SECRET_LOOKUP_TIMEOUT_SECS * G_TIME_SPAN_SECOND;
+    while (!tc->priv->password) {
+        if (!g_cond_wait_until (&tc->priv->passwordCond, &tc->priv->passwordMutex, timeout)) {
```
Comment (TingPing) [inline src/trg-client.c]: You didn't actually make this non-blocking.. you just changed the block.

**Comment by TingPing** [comment]
How it would be designed non-blocking is ideally the UI would change to show a loading state or such and it happens in the background.

**Comment by snaunton** [comment]
No worries. I shouldn't code when I am sick :(. Redesigned to load with prefs. So, the passwords are in memory...

**Comment by snaunton** [comment]
Back to using secret_password_lookup_(nonpageable_)sync in get_curl, but have added a timeout (currently hard coded to three seconds), using GCancellable, and status/error messages for user. Note that libsecret returns NULL instantly when gnome-keyring-daemon is stopped (at least 0.18.5 on debian testing/buster anyway), so the timeout is theoretical at present. The same idea works in different circumstances though.

Added an entry dialog for the password which can only set it, IOW the password is only ever loaded in get_curl.

FWIW, I couldn't find the password when dumping memory, so yay I guess.

Hopefully I have made libsecret not suck too much...? :)

File: src/trg-client.c [hunk 3]
```diff
@@ -517,6 +561,50 @@ static trg_tls *get_tls(TrgClient *tc) {
...
+    TrgClientPrivate *priv = tc->priv;
+    const gchar *profile_uuid = trg_prefs_get_profile_uuid(priv->prefs);
+    GError *error = NULL;
+
+    /* Inform anyone who wants to know that we are fetching password */
+    trg_client_fetch_password_emit_signal(tc);
+
+    /* Use GCancellable and g_timeout_add to create a timeout for password lookup */                                      
+    GCancellable *cancellable = g_cancellable_new();
+    guint timeout_func = g_timeout_add (3000, get_secret_password_timeout_cb, cancellable);
```
Comment (TingPing) [inline src/trg-client.c]: This will never run though, the thread is blocked.

File: src/trg-client.c [hunk 4]
```diff
@@ -517,6 +561,50 @@ static trg_tls *get_tls(TrgClient *tc) {
...
+                                                               &error,
+                                                               TRG_PREFS_KEY_PROFILE_UUID,
+                                                               profile_uuid,
+                                                               NULL);
+
+    if(!g_cancellable_is_cancelled(cancellable))
+        g_source_remove(timeout_func);
+
+    if(!password || error ) {
+       if(error)
```
Comment (TingPing) [inline src/trg-client.c]: `g_clear_error(&error);`

File: src/trg-preferences-dialog.c [hunk 5]
```diff
@@ -592,6 +626,73 @@ static void add_profile_cb(GtkWidget * w, gpointer data)
...
+
+    if(!secret_password_store_finish (result, NULL)) {
+        GtkWidget *dialog = gtk_message_dialog_new(GTK_WINDOW(win),
+                                                   GTK_DIALOG_MODAL,
+                                                   GTK_MESSAGE_ERROR,
+                                                   GTK_BUTTONS_OK,
+                                                   "%s", _("Failed to save password to keychain"));
+
+        gtk_window_set_title(GTK_WINDOW(dialog), _("Error"));
+        gtk_dialog_run(GTK_DIALOG(dialog));
```
Comment (TingPing) [inline src/trg-preferences-dialog.c]: Never use `gtk_dialog_run()`. Show the dialog and connect to the `response` signal.

File: src/trg-secret.c [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    g_source_set_callback(source, trg_secret_get_password_timeout, cancellable, NULL);
+    g_source_attach(source, sync->context);
+    g_source_unref(source);
+
+    secret_password_lookup(TRG_SECRET_SCHEMA, cancellable,
+                           trg_secret_sync_on_result, sync,
+                           TRG_PREFS_KEY_PROFILE_UUID, uuid,
+                           NULL);
+
+    g_main_loop_run(sync->loop);
```
Comment (TingPing) [inline src/trg-secret.c]: Sorry to be a pest but boy this is overkill. Use [GTask](https://developer.gnome.org/gio/stable/GTask.html) to run the sync variant in its own thread. Then your logic from before should work fine. The Cancellable will trigger cross thread and the timer won't be blocked.

File: src/trg-secret.c [hunk 7]
```diff
@@ -96,40 +62,47 @@ trg_secret_get_password_timeout(gpointer userdata)
...
-                           TRG_PREFS_KEY_PROFILE_UUID, uuid,
-                           NULL);
+    guint timeout_source_id = g_timeout_add (timeout, 
+                                             trg_secret_get_password_timeout,
+                                             task_data->cancellable);
 
-    g_main_loop_run(sync->loop);
+    GTask *task = g_task_new(NULL, NULL, NULL, NULL);
+    g_task_set_task_data(task, task_data, NULL);
+    g_task_run_in_thread_sync(task, trg_secret_get_password_task);
```
Comment (TingPing) [inline src/trg-secret.c]: Remove the `_sync`

**Comment by snaunton** [comment]
Exposing myself to looking like a big dummy for this, by why remove _sync?

Just not using sync will not work, as is.

I am a bit confused here. My idea behind using libsecret is also to minimize the amount of time the password is in memory in plain text*, thus calling in get_curl(), rather than in trg_client_populate_with_settings().

In either case, the app is going to have to wait for libsecret to return - the password is needed to make the connection.

So, to use async we could get the password and call dispatch_async etc from the callback, but this means that password is left in memory longer, is harder to determine when to free it, requires more complex code etc.

The password is only (seems to be) got three times, and only when the connection is made, and perhaps most importantly the threads are IO blocked by curl anyway. So is using sync for this that much of a big deal? If we cannot get the password the connection is going to fail anyway. If we can get the password it is quick.

However, having written all that, you certainly know the code base (and gtk, glib, programming in general probably) better than me, so I assume I must be missing something here...?

Sorry, for the amount of text :).

*Sure, there are loads of arguments against worrying about this, but I'd rather have libsecret (and other stuff) or the user's own security choices be the weaker link than this.

**Comment by TingPing** [comment]
> Exposing myself to looking like a big dummy for this, by why remove _sync?

Because `_sync` blocks the main thread, locking up everything.

> In either case, the app is going to have to wait for libsecret to return - the password is needed to make the connection.

While true the user needs to wait locking up the application isn't the way to do it. That might prompt your shell to ask to forcefully kill the application, stop responding to clicks and keys like preventing the user from closing the window, etc. It is bad UX and ideally you show something like a loading screen or at least a status message.

> So, to use async we could get the password and call dispatch_async etc from the callback, but this means that password is left in memory longer

If it is marked as secure memory, which libsecret does afaik, that doesn't matter. Feel free to correct me if I am wrong there.

>  requires more complex code 

Can't argue that.

> perhaps most importantly the threads are IO blocked by curl anyway

If curl usage is actually blocking the UI then that is a bug.

>  If we can get the password it is quick.

IO isn't "quick" it is "an undermined amount of time between now and never". Hoping it is fast is just bad design.

> However, having written all that, you certainly know the code base (and gtk, glib, programming in general probably) better than me, so I assume I must be missing something here...?

I'm actually not super familiar with this codebase, I sort of fell into being a maintainer because nobody else was, and I actually started [rewriting it](https://github.com/TingPing/transmission-remote-gnome) because I wasn't a huge fan. I am familiar with Gtk though and not blocking rendering is a pretty simple rule to live by.

**Comment by snaunton** [comment]
> Because _sync blocks the main thread, locking up everything.

Not in this case. g_task_run_in_thread_sync(task, trg_secret_get_password_task); is running in a different thread to the UI.

As a test I added while(1) to trg-secret.c -> trg_secret_get_password()

`    g_task_run_in_thread_sync(task, trg_secret_get_password_task);

    while(1);

    if(!g_cancellable_is_cancelled(task_data->cancellable))
        g_source_remove(timeout_source_id);
`

This blocks the thread, but message shown in statusbar is "Fetching password..." and the UI is completely operational (well, in so far as it should be operational without a connection), so the below is not applicable:

> While true the user needs to wait locking up the application isn't the way to do it. That might prompt your shell to ask to forcefully kill the application, stop responding to clicks and keys like preventing the user from closing the window, etc. It is bad UX and ideally you show something like a loading screen or at least a status message.

> If it is marked as secure memory, which libsecret does afaik, that doesn't matter. Correct me if I am wrong there.

AFAIK, libsecret stores passwords in memory encrypted (might be on disk... ¯\_(ツ)_/¯, but encypted regardless), but once they are are returned by a lookup call to the calling process they are in plain text. They can be stored in non-pageable memory, but they are still in plain text. I've checked by dumping the memory and scanning for my password. It is there in plain text. secret_password_free() simply overwrites the memory before freeing it (and can free non-pageable memory).

> If curl usage is actually blocking the UI then that is a bug.

Curl is running in separate threads to the GUI thread. However, get_curl() has a mutex, so it halts all the client threads (the threads which communicate with transmission-daemon). trg_secret_get_password is called within the mutex so can potentially halt all the other client threads.

I think this is where I was getting confused. I thought you were concerned about blocking the other client threads.

However, WRT above I'm can't think of a scenario where this would actually be a problem as the password is only got when the connection is being made, so if there are any other client threads at this time (I don't think there are) I am not sure what they would be doing.

TL;DR, curl is not blocking the UI. Neither is the call to libsecret.

I hope all this makes sense.

On a side note... thanks for being a hard on this. It has really made me think about the code rather than just hacking something together, and forced me to learn a fair bit about theads and tasks in glib/gtk which I did know before.
